src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/SignatureHelper.cs

```diff
@@ -63,19 +64,83 @@ internal static BlobBuilder MethodSignatureEncoder(ModuleBuilderImpl module, Typ
             {
                 foreach (Type parameter in parameters)
                 {
-                    WriteSignatureForType(parEncoder.AddParameter().Type(), parameter, module);
+                    WriteSignatureForType(parEncoder.AddParameter().Type(parameter.IsByRef), parameter, module);
                 }
             }
 
             return methodSignature;
         }
 
         private static void WriteSignatureForType(SignatureTypeEncoder signature, Type type, ModuleBuilderImpl module)
+        {
+            if (type.IsArray)
+            {
+                Type elementType = type.GetElementType()!;
+                int rank = type.GetArrayRank();
+                if (rank == 1)
+                {
+                    WriteSignatureForType(signature.SZArray(), elementType, module);
+                }
+                else
+                {
+                    signature.Array(out SignatureTypeEncoder elTypeSignature, out ArrayShapeEncoder arrayEncoder);
+                    WriteSimpleSignature(elTypeSignature, elementType, module);
+                    arrayEncoder.Shape(type.GetArrayRank(), ImmutableArray.Create<int>(), ImmutableArray.Create<int>(new int[rank]));
```

Comment:
I suspect that this does not work. Do you have a test for multi-dimensional arrays that hits this path? 

(The helper classes for signature encoding are not very help. One still very much needs to know how the signature is encoded while using them. To me, it would be more obvious what's going on we wrote the signature into the BlobBuilder ourselves.)